cognite/client/utils/_priority_tpe.py

```diff
@@ -30,32 +30,49 @@
 """
 from __future__ import annotations
 
-import atexit
 import inspect
 import itertools
 import sys
 import threading
 import weakref
-from concurrent.futures.thread import ThreadPoolExecutor, _base, _python_exit, _threads_queues, _WorkItem
+from concurrent.futures.thread import ThreadPoolExecutor, _base, _WorkItem
 from queue import PriorityQueue
-from threading import Lock
 
 NULL_ENTRY = (sys.maxsize, None, _WorkItem(None, None, (), {}))
+# Keep a separate weakref. dict (thread queue) from 'concurrent.futures.thread._thread_queue' in order
+# for vanilla ThreadPoolExecutor to work in tandem with PriorityThreadPoolExecutor - or more correctly,
+# get its threads cleaned up correctly.
+_THREADS_QUEUES = weakref.WeakKeyDictionary()
 _SHUTDOWN = False
+# Lock that ensures that new workers are not created while the interpreter is
+# shutting down. Must be held while mutating _THREADS_QUEUES and _SHUTDOWN.
+_GLOBAL_SHUTDOWN_LOCK = threading.Lock()
 
 
 def python_exit():
     global _SHUTDOWN
-    _SHUTDOWN = True
-    items = list(_threads_queues.items())
+    with _GLOBAL_SHUTDOWN_LOCK:
+        _SHUTDOWN = True
+    items = list(_THREADS_QUEUES.items())
     for thread, queue in items:
         queue.put(NULL_ENTRY)
     for thread, queue in items:
         thread.join()
 
 
-atexit.unregister(_python_exit)
-atexit.register(python_exit)
+# We should register an at-exit-handler to clean up our threads gracefully at interpreter shutdown.
```

Comment:
I don't really think there is a need for "cleaning up the threads gracefully". Essentially you're just waiting for work to finish, and then tossing the results out, so the only thing you achieve is slower shutdown of the interpreter. Unless I'm missing something here, I think we can remove the atexit hook altogether.

Reply:
With current tasks, yes, however, that gives us a "dependency" on the future use of the PTPE to never have tasks with side effects; not something I would want to perpetually make sure is the case.

Reply:
I've gone down the deep rabbit hole here and I really think this implementation is the way way to do it. Separate thread pools, separate thread-safe queues, and separate cleanup functions registered that take the python version into account! 🍀

Reply:
Fair point - let's go for this approach then! 🤝

Reply:
Lol YAGNI, learned something new today 😂 Btw, if you really want it removed why don’t we also remove from the standard library by monkeypatching? 😜 I feel good knowing we mimic the exact same approach as some code written by people way smarter than myself 😂

Reply:
Also; the code do crash without it 😉